source/03-components/table/table--sortable.es6.ts

```diff
@@ -0,0 +1,295 @@
+      : 'unsorted'
+  }`;
+  const headerButtonLabel = `Click to sort by ${headerName} in ${
+    sortedAscending ? DESCENDING : ASCENDING
+  } order.`;
+  header.setAttribute('aria-label', headerLabel);
+  header.querySelector(SORT_BUTTON)?.setAttribute('title', headerButtonLabel);
+};
+
+/**
+ * Remove the aria-sort attribute on the given header element, and reset the
+ * label and button icon.
+ * @param {HTMLTableCellElement} header
+ */
+const unsetSort = (header: HTMLTableCellElement): void => {
+  header.removeAttribute(SORTED);
+  updateSortLabel(header);
+};
+
+/**
+ * Sort rows either ascending or descending, based on a given header’s
+ * aria-sort attribute.
+ * @param {HTMLTableCellElement} header
+ * @param {boolean} isAscending
+ * @return {boolean}
+ */
+const sortRows = (
+  header: HTMLTableCellElement,
+  isAscending: boolean
+): boolean => {
+  header.setAttribute(SORTED, isAscending === true ? DESCENDING : ASCENDING);
+  updateSortLabel(header);
+
+  const table = header.closest(TABLE);
+  const headerRow = header.parentNode;
+
+  if (!table || !headerRow) {
+    return false;
+  }
+
+  const tbody = table.querySelector('tbody') as HTMLTableSectionElement;
+  const siblingHeaders = headerRow.children as HTMLCollection;
+
+  Array.from(tbody.querySelectorAll('tr'))
+    .sort(
+      compareFunction(Array.from(siblingHeaders).indexOf(header), !isAscending)
+    )
+    .forEach(tr => tbody.appendChild(tr));
+
+  return true;
+};
+
+/**
+ * Update the live region immediately following the table whenever the sort
+ * changes.
+ * @param {HTMLTableElement} table
+ * @param {HTMLTableCellElement} sortedHeader
+ */
+const updateLiveRegion = (
+  table: HTMLTableElement,
+  sortedHeader: HTMLTableCellElement
+): void => {
+  const caption = table.querySelector('caption')?.innerText;
+  const sortedAscending = sortedHeader.getAttribute(SORTED) === ASCENDING;
+  const headerLabel = sortedHeader.innerText;
+  const liveRegion = table.nextElementSibling as HTMLElement;
+
+  if (liveRegion && liveRegion.matches(ANNOUNCEMENT_REGION)) {
+    let tableName = '';
+    if (caption) {
+      tableName = ` named “${caption}”`;
+    }
+
+    const sortAnnouncement = `The table${tableName} is now sorted by ${headerLabel} in ${
+      sortedAscending ? ASCENDING : DESCENDING
+    } order.`;
+    liveRegion.innerText = sortAnnouncement;
+  } else {
+    // eslint-disable-next-line no-console
+    console.error(
+      'Table containing a sortable column header is not followed by an aria-live region.'
+    );
+  }
+};
+
+/**
+ * Toggle a header’s sort state, optionally providing a target state. If no
+ * state is provided, the current state will be toggled (from false to true,
+ * and vice-versa).
+ *
+ * @param {HTMLTableCellElement} header
+ * @param {boolean?} isAscending
+ */
+const toggleSort = (
+  header: HTMLTableCellElement,
+  isAscending: boolean
+): void => {
+  const table = header.closest(TABLE) as HTMLTableElement;
```

Comment:
Suggest removing the casting and instead doing a check alongside `!table` down below:
```ts
const table = header.closest(TABLE);
...
if (!table || !(table instanceof HTMLTableElement)) {
  ...
}
```